src/relax/transform/lambda_lift.cc

```diff
@@ -336,176 +266,235 @@ class LambdaLifter : public ExprMutator {
+    }
+
+    return ExprMutator::VisitExpr_(call.get());
+  }
+
+  Expr VisitExpr_(const VarNode* op) override {
+    auto var = GetRef<Var>(op);
+    if (auto it = rebind_map_.find(var); it != rebind_map_.end()) {
+      return it->second;
+    }
+    return ExprMutator::VisitExpr_(op);
+  }
+
+  bool IsClosure(Expr val) {
+    static int depth = -1;
+    struct Context {
+      explicit Context(int* ptr) : ptr(ptr) { (*ptr)++; }
+      ~Context() { (*ptr)--; }
+      int* ptr;
+    } context(&depth);
+
+    if (auto opt_var = val.as<Var>()) {
+      if (closures_.count(opt_var.value())) {
+        return true;
+      }
+      if (auto bound_value = builder_->LookupBinding(opt_var.value())) {
+        val = bound_value.value();
+      }
+    }
+
+    if (const auto* call_node = val.as<relax::CallNode>()) {
       // recursive call
       auto op = call_node->op;
-      if (make_closure_op_ == op) {
+      if (auto local_var = op.as<Var>()) {
+        return IsClosure(local_var.value());
+      } else if (auto global_var = op.as<GlobalVar>()) {
+        return IsClosure(global_var.value());
+      } else {
+        return make_closure_op_ == op;
+      }
+
+    } else if (const auto* global_var = val.as<GlobalVarNode>()) {
+      if (closures_.count(GetRef<GlobalVar>(global_var))) {
         return true;
       }
-      if (const auto* lv = op.as<VarNode>()) {
-        return HasClosure(GetRef<Var>(lv));
+      IRModule ctx_mod = builder_->GetContextIRModule();
+      ICHECK(ctx_mod->functions.size() > 0);
+      BaseFunc func = ctx_mod->Lookup(GetRef<GlobalVar>(global_var));
+      const auto* func_node = func.as<FunctionNode>();
+      if (func_node) {
+        return IsClosure(func_node->body);
+      } else {
+        return false;
       }
+
+    } else if (const auto* func_node = val.as<FunctionNode>()) {
+      return IsClosure(func_node->body);
+
+    } else if (const auto* seq_node = val.as<SeqExprNode>()) {
+      return IsClosure(seq_node->body);
+
+    } else {
+      return false;
     }
-    return false;
   }
 
-  bool IsClosure(const Array<Var>& captured_vars) { return captured_vars.size() > 0; }
-
   IRModule Lift() {
     auto glob_funcs = mod_->functions;
-    for (auto pair : glob_funcs) {
-      if (auto* n = pair.second.as<FunctionNode>()) {
-        auto func = GetRef<Function>(n);
-        func = Function(func->params, VisitExpr(func->body), func->ret_struct_info, func->is_pure,
-                        func->attrs);
-        builder_->UpdateFunction(pair.first, func);
+    for (auto [gvar, base_func] : glob_funcs) {
+      if (auto opt = base_func.as<Function>()) {
+        // Must visit the function itself, and not just the function
+        // body, to ensure that EraseToWellDefined recognized symbolic
+        // variables that are exposed by the function signature.
+        auto func = Downcast<Function>(VisitExpr(opt.value()));
+        builder_->UpdateFunction(gvar, func);
```

Comment:
Is this the only change needed to handle symbolic vars, fundamentally? I ask because I didn't see any logic specific to symbolic vars above.

Reply:
Primarily, yes.  Most of the other changes are to allow this specific change to work.  Previously, the mutator could assume that every `FunctionNode` encountered would be a lambda to be lifted out, but now it could be a top-level function as well.